File: Services/Logger.cs
using System;
using System.Diagnostics;
using System.IO;

namespace AnimeWallpaper.Services;

public static class Logger
{
    private static readonly object _lock = new();
    private static string? _configuredApiKey;

    public static string LogDirectory => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "AnimeWallpaper",
        "logs"
    );

    public static void SetApiKeyToMask(string? apiKey)
    {
        _configuredApiKey = apiKey;
    }

    public static void Info(string message) => Log("INFO", message);
    public static void Warn(string message) => Log("WARN", message);
    public static void Error(string message, Exception? ex = null)
    {
        var fullMessage = ex == null ? message : $"{message} | Exception: {ex.Message}\n{ex.StackTrace}";
        Log("ERROR", fullMessage);
    }

    private static void Log(string level, string message)
    {
        try
        {
            if (!string.IsNullOrEmpty(_configuredApiKey) && _configuredApiKey.Length >= 4)
            {
                message = message.Replace(_configuredApiKey, "***API_KEY_HIDDEN***");
            }

            var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [{level}] {message}";
            Debug.WriteLine(line);

            lock (_lock)
            {
                if (!Directory.Exists(LogDirectory))
                {
                    Directory.CreateDirectory(LogDirectory);
                }

                var logFilePath = Path.Combine(LogDirectory, $"app-{DateTime.Now:yyyyMMdd}.log");
                File.AppendAllText(logFilePath, line + Environment.NewLine);
            }
        }
        catch
        {
            // Logging should never throw and crash the application
        }
    }

    public static void OpenLogFolder()
    {
        try
        {
            if (!Directory.Exists(LogDirectory))
            {
                Directory.CreateDirectory(LogDirectory);
            }
            Process.Start(new ProcessStartInfo
            {
                FileName = LogDirectory,
                UseShellExecute = true
            });
        }
        catch (Exception ex)
        {
            Error("Failed to open log directory", ex);
        }
    }
}

File: macos/Sources/Services/Logger.swift
import Foundation
import AppKit

public final class Logger: @unchecked Sendable {
    public static let shared = Logger()
    private let lock = NSLock()
    private var configuredApiKey: String?

    public var logDirectory: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("AnimeWallpaper/logs")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    public func setApiKeyToMask(_ key: String?) {
        lock.lock()
        defer { lock.unlock() }
        configuredApiKey = key
    }

    public static func info(_ message: String) {
        shared.log(level: "INFO", message: message)
    }

    public static func warn(_ message: String) {
        shared.log(level: "WARN", message: message)
    }

    public static func error(_ message: String, error: Error? = nil) {
        let fullMessage = error == nil ? message : "\(message) | Error: \(error!.localizedDescription)"
        shared.log(level: "ERROR", message: fullMessage)
    }

    private func log(level: String, message: String) {
        var sanitized = message
        lock.lock()
        if let key = configuredApiKey, key.count >= 4 {
            sanitized = sanitized.replacingOccurrences(of: key, with: "***API_KEY_HIDDEN***")
        }
        lock.unlock()

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let timestamp = formatter.string(from: Date())
        let line = "\(timestamp) [\(level)] \(sanitized)\n"

        print(line, terminator: "")

        lock.lock()
        defer { lock.unlock() }
        let dayFormatter = DateFormatter()
        dayFormatter.dateFormat = "yyyyMMdd"
        let logFile = logDirectory.appendingPathComponent("app-\(dayFormatter.string(from: Date())).log")

        if let data = line.data(using: .utf8) {
            if FileManager.default.fileExists(atPath: logFile.path) {
                if let fileHandle = try? FileHandle(forWritingTo: logFile) {
                    fileHandle.seekToEndOfFile()
                    fileHandle.write(data)
                    try? fileHandle.close()
                }
            } else {
                try? data.write(to: logFile, options: .atomic)
            }
        }
    }

    public static func openLogFolder() {
        NSWorkspace.shared.open(shared.logDirectory)
    }
}
